File: backend/puzzles/engine/rules/lightup.py
from typing import Optional

from puzzles.engine.games import PuzzleEngineBase
from puzzles.engine.rules.base import RuleDefinition, ValidationResult


def no_bulbs_lighting_each_other() -> RuleDefinition:
    """
    Generate a rule to validate that no two bulbs can see each other in Light Up.

    In Light Up, bulbs shine light in all four cardinal directions until they hit
    a wall (black cell). No two bulbs should be able to see each other.

    :return: A rule function that validates bulb visibility constraints
    """

    def rule(engine: PuzzleEngineBase) -> Optional[ValidationResult]:
        from puzzles.engine.games.lightup import LightupCellStates

        state = engine.board_state
        rows, cols = engine.rows, engine.cols

        # Wall values from LightupCellStates enum
        wall_values = {
            LightupCellStates.Wall0,
            LightupCellStates.Wall1,
            LightupCellStates.Wall2,
            LightupCellStates.Wall3,
            LightupCellStates.Wall4,
            LightupCellStates.WallBlack,
        }

        # Find all bulbs
        bulbs = []
        for r in range(rows):
            for c in range(cols):
                if state[r * cols + c] == LightupCellStates.Bulb:
                    bulbs.append((r, c))

        violating_bulbs_set = set()

        # Check each bulb against all others
        for i, (r1, c1) in enumerate(bulbs):
            for j, (r2, c2) in enumerate(bulbs):
                if i >= j:  # Avoid checking same pair twice
                    continue

                # Check if bulbs can see each other (same row or column, no walls between)
                if r1 == r2:  # Same row
                    min_c, max_c = min(c1, c2), max(c1, c2)
                    can_see = True
                    for c in range(min_c + 1, max_c):
                        if state[r1 * cols + c] in wall_values:
                            can_see = False
                            break
                    if can_see:
                        violating_bulbs_set.add((r1, c1))
                        violating_bulbs_set.add((r2, c2))

                elif c1 == c2:  # Same column
                    min_r, max_r = min(r1, r2), max(r1, r2)
                    can_see = True
                    for r in range(min_r + 1, max_r):
                        if state[r * cols + c1] in wall_values:
                            can_see = False
                            break
                    if can_see:
                        violating_bulbs_set.add((r1, c1))
                        violating_bulbs_set.add((r2, c2))

        if not violating_bulbs_set:
            return None

        violating_bulbs_list = [{"row": r, "col": c} for r, c in violating_bulbs_set]

        return ValidationResult(
            locations=violating_bulbs_list,
            rule_type="bulb_intersection_violation",
        )

    return RuleDefinition(rule)


def validate_numbered_wall_constraints() -> RuleDefinition:
    """
    Generate a rule to validate that numbered walls in Light Up have the correct
    number of adjacent bulbs.

    In Light Up, walls with numbers (0-4) must have exactly that many bulbs
    in the 4 adjacent cells (not diagonals).

    :return: A rule function that validates numbered wall constraints
    """

    def rule(engine: PuzzleEngineBase) -> Optional[ValidationResult]:
        from puzzles.engine.games.lightup import LightupCellStates

        state = engine.board_state
        rows, cols = engine.rows, engine.cols

        # Only 4 cardinal directions for Light Up wall constraints
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        violating_walls = []

        for r in range(rows):
            for c in range(cols):
                cell_value = state[r * cols + c]

                # Check if this is a numbered wall (0-4)
                if cell_value not in [
                    LightupCellStates.Wall0,
                    LightupCellStates.Wall1,
                    LightupCellStates.Wall2,
                    LightupCellStates.Wall3,
                    LightupCellStates.Wall4,
                ]:
                    continue

                # Get the required number of bulbs
                required_bulbs = cell_value  # Wall0 = 0, Wall1 = 1, etc.

                # Count adjacent bulbs
                adjacent_bulbs = 0
                for dr, dc in directions:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < rows and 0 <= nc < cols:
                        if state[nr * cols + nc] == LightupCellStates.Bulb:
                            adjacent_bulbs += 1

                # Check if constraint is violated (too many bulbs)
                if adjacent_bulbs > required_bulbs:
                    violating_walls.append({"row": r, "col": c})

        if not violating_walls:
            return None

        return ValidationResult(locations=violating_walls, rule_type="numbered_wall_constraint_violated")

    return RuleDefinition(rule)

File: frontend/src/services/game/engines/rules/lightup.ts
import type {
  PuzzleEngine,
  RuleDefinition,
  RuleViolation,
  RuleViolationCell,
} from "@/services/game/engines/PuzzleEngine.ts";
import { LightupCell } from "@/services/game/engines/translator.ts";

/* ───────────────────────── helper ───────────────────────── */
const inBounds = (r: number, c: number, rows: number, cols: number) => r >= 0 && r < rows && c >= 0 && c < cols;

export function no_bulbs_lighting_each_other(): RuleDefinition {
  function rule(engine: PuzzleEngine): RuleViolation | null {
    const state = engine.board_state;
    const rows = engine.definition.rows;
    const cols = engine.definition.cols;

    const wallValues = new Set<number>([
      LightupCell.WALL_0,
      LightupCell.WALL_1,
      LightupCell.WALL_2,
      LightupCell.WALL_3,
      LightupCell.WALL_4,
      LightupCell.WALL_NO_CONSTRAINT,
    ]);

    const bulbs: [number, number][] = [];
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        if (state[r][c] === LightupCell.BULB) bulbs.push([r, c]);
      }
    }

    /* cast rays from each bulb */
    const hit = new Set<string>();
    const dirs = [
      [-1, 0],
      [1, 0],
      [0, -1],
      [0, 1],
    ];

    for (const [r, c] of bulbs) {
      for (const [dr, dc] of dirs) {
        let nr = r + dr,
          nc = c + dc;
        while (inBounds(nr, nc, rows, cols) && !wallValues.has(state[nr][nc] as number)) {
          if (state[nr][nc] === LightupCell.BULB) {
            hit.add(`${r},${c}`);
            hit.add(`${nr},${nc}`);
            break; // stop in this direction
          }
          nr += dr;
          nc += dc;
        }
      }
    }

    if (hit.size === 0) return null;

    const locations: RuleViolationCell[] = Array.from(hit).map((key) => {
      const [row, col] = key.split(",").map(Number);
      return { row, col };
    });

    return {
      locations,
      rule_name: "bulb_intersection_violation",
    };
  }

  return { rule_func: rule };
}

/* ======================================================================
 * 2. Numbered walls must not have *more* adjacent bulbs than the number.
 * ==================================================================== */
export function validate_numbered_wall_constraints(): RuleDefinition {
  function rule(engine: PuzzleEngine): RuleViolation | null {
    const state = engine.board_state;
    const rows = engine.definition.rows;
    const cols = engine.definition.cols;

    const dirs = [
      [-1, 0],
      [1, 0],
      [0, -1],
      [0, 1],
    ];

    const numberedWalls = [
      LightupCell.WALL_0,
      LightupCell.WALL_1,
      LightupCell.WALL_2,
      LightupCell.WALL_3,
      LightupCell.WALL_4,
    ];

    const violating: RuleViolationCell[] = [];

    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const cell = state[r][c];
        if (!numberedWalls.includes(cell as number)) continue;

        const required = cell as number; // Wall0 ⇒ 0, Wall1 ⇒ 1, …

        let bulbs = 0;
        for (const [dr, dc] of dirs) {
          const nr = r + dr,
            nc = c + dc;
          if (inBounds(nr, nc, rows, cols) && state[nr][nc] === LightupCell.BULB) {
            bulbs++;
          }
        }

        if (bulbs > required) violating.push({ row: r, col: c });
      }
    }

    return violating.length ? { locations: violating, rule_name: "numbered_wall_constraint_violated" } : null;
  }

  return { rule_func: rule };
}
